### ISSUE TITLE
Expose an instance local schema version

### ISSUE THREAD

**Issue by Totktonada** [tarantool/tarantool#6544]
There is internal `box.internal.schema_version()`, but there is no public function of the same kind.

It is useful for caching of some database schema information, possibly transformed for easier / faster access (see https://github.com/tarantool/ddl/issues/69).

Sometimes it is much easier to compare schema versions rather than set triggers on system spaces. Especially, when you attempt to do so in a module (not an application) in a transparent way (without 'call a special module function to set triggers for updating caches, but do this only after full box loading').

<details>
<summary>What is wrong with using triggers to update a database schema cache in a module?</summary>

* If box is **fully** loaded[^1], create the initial cache and set on_replace trigger on `_space`, `_index` and `_collation` to update the cache.
* If box is not configured and tarantool is 2.*, set on_schema_init trigger to create the initial cache and set on_replace.
* If box is not configured and tarantool is 1.10, spawn a fiber to check `_space`, `_index`, `_collection` appearance (and hope that [nobody](https://github.com/tarantool/cartridge/blob/1d477838b097ccabd9ae7e0242951be12f4c8036/cartridge/hotreload.lua#L144-L174) will kill it).
* If box is partially configured, I don't know. Whether on_schema_init will call the trigger if it is set after appearance of necessary space, but before full box loading (seems no)? Maybe we'll need to fallback to the fiber solution.

[^1]: It is not easy to determine. No, `type(box.cfg)` is not the option. At least, it becomes table before **full** box load and there is no way to distinguish this transient state from full box load (except catching box.cfg exitting, but that's application code).
</details>

We should provide formal rules, when the schema version is increased (I know, on DDL, but what exactly is DDL in context of tarantool?). (Maybe this is topic for another issue.)

Things to consider:

1. A return value when box is not configured.
2. A return value during recovery (database loading).
3. A return value in a transaction, which should not see effect of some recent DDL operation.

It should be okay to don't support those three cases at all and return an error (or nil, or zero; I don't know). However I'm not sure about transactions (the third point).

<details>
<summary>Whether to support calling box.schema_version() in a transaction?</summary>

We don't need to collect `_space`, `_index` and `_collation` in a transaction, because those are memtx spaces, so reading from them does not yield. Constructing the cache unlikely requires yielding, because I doubt that someone need something aside constructing Lua tables here. But if we're in a module, we don't control, whether our module function (that uses the schema cache and may trigger its refreshing and so calling `box.schema_version()`) is called in a transaction or not. So `box.schema_version()` should either return correct value for current transaction or return an error (and we'll declare that calling it in a transaction is not supported).

Don't know. I would prefer transactional `box.schema_version()`, because the function that uses schema cache may be, say, just 'whether given space has JSON path indices' or something like this. It would be strange to forbid calling such function is a transaction.

If we'll going to raise an error with idea to support transactions in a future, it should be declared explicitly, because affects how a module code will use the function (see examples below).
</details>

<details>
<summary>Sketchy example of usage</summary>

I wrote it for myself to generate points above (it is how I'm thinking), but it is may be useful to better understand my wishes and doubts.

The best way to use this example is write your own and check, whether it handles all corner cases, which are handled here.

(I didn't run it, so it may contain stupid mistakes.)

```lua
-- Global cache object.
local global_cache = {
    -- <..some data..>,
    -- schema_version = <number> (<nil> initially),
}

local function box_not_configured_error()
    error("Unable to <..what we're doing here..>: ' ..
        'box is not configured or is in process of loading")
end

-- Generate the schema cache object.
--
-- It doesn't touch the global cache object.
local function generate_cache(box_schema_version)
    local cache = {}
    <..collect data from _space, _index, _collation and write it to cache..>
    cache.schema_version = box_schema_version
    return cache
end

-- Depending of whether the global cache is actual and whether
-- we're in a transaction, return appropriate cache object.
--
-- Assume box.schema_version() is transactional.
local function choose_cache()
    -- XXX: If box.schema_version will raise an error
    -- before box configuration on accessing the 'schema_version'
    -- field (not only on the call of the function), we need a bit
    -- more complex code here:
    --
    --  | -- Accessing box.schema_version may raise an error, so
    --  | -- we need this extra wrapping function.
    --  | local ok, box_schema_version = pcall(function()
    --  |     return box.schema_version()
    --  | end)
    local ok, box_schema_version = pcall(box.schema_version)
    if not ok then
        return box_not_configured_error()
    end

    assert(box_schema_version ~= nil)
    if global_cache.schema_version == box_schema_version then
        return global_cache
    end

    -- We can see not a latest schema in a transaction, so
    -- store the schema cache only if it is newer than the
    -- already stored one.
    local cache = generate_cache(box_schema_version)
    if global_cache.schema_version < cache.schema_version then
        global_cache = cache
    end
    return cache
end

-- Depending of whether the global cache is actual and whether
-- we're in a transaction, return appropriate cache object.
--
-- Assume box.schema_version() raises an error in a transaction.
--
-- There is also version of this code that will works in both
-- cases (error in a txn, corrent version in a txn), see comments.
-- It will be usable, if initial box.schema_version() version will
-- raise an error in a transaction, but a future tarantool version
-- may add support of this case.
local function choose_cache()
    -- XXX: If box.schema_version will raise an error
    -- before box configuration on accessing the 'schema_version'
    -- field (not only on the call of the function), we need a bit
    -- more complex code here:
    --
    --  | -- Accessing box.schema_version may raise an error, so
    --  | -- we need this extra wrapping function.
    --  | local ok, box_schema_version = pcall(function()
    --  |     return box.schema_version()
    --  | end)
    local ok, box_schema_version = pcall(box.schema_version)
    if not ok then
        box_schema_version = nil

        -- We call box.is_in_txn() only if box.schema_version()
        -- fails. It is a little optimization for the general
        -- 'not in txn' case.

        -- XXX: Whether is is possible that the box is not
        -- **fully** loaded, but box.is_in_txn() succeeds?
        -- I hope, no: I don't know how to distinguish this
        -- case from when box is fully configured.
        --
        -- But if it is so, we definitely need a function
        -- from box, which will say us, whether box is **fully**
        -- loaded. BTW, it would simplify the code.

        -- Accessing box.is_in_txn may raise an error, so
        -- we need this extra wrapping function.
        local ok, is_in_txn = pcall(function()
            return box.is_in_txn()
        end)

        -- Whether box.schema_version() fails because of
        -- unconfigured box or because we're in a transaction?
        if not ok or not is_in_txn then
            return box_not_configured_error()
        end

        -- Here we're in a transaction.
        --
        -- No way to determine a schema version that corresponds
        -- to given database state.
        return generate_cache()
    end

    -- We're not in a transaction here (but see below).
    --
    -- XXX: If we assume that box.schema_version() may start
    -- to work in a transaction in future tarantool versions,
    -- the code should be the following:
    --
    --  | assert(box_schema_version ~= nil)
    --  | if global_cache.schema_version == box_schema_version then
    --  |     return global_cache
    --  | end
    --  |
    --  | We can see not a latest schema in a transaction, so
    --  | store the schema cache only if it is newer than the
    --  | already stored one.
    --  | local cache = generate_cache(box_schema_version)
    --  | if global_cache.schema_version < cache.schema_version then
    --  |     global_cache = cache
    --  | end
    assert(box_schema_version ~= nil)
    if global_cache.schema_version ~= box_schema_version then
        global_cache = generate_cache(box_schema_version)
    end

    return global_cache
end

local function a_func_that_uses_cache()
    local cache = choose_cache()
    <..access cache..>
end
```

What is definitely not handled in this code:

* Access from different fibers (see the [locked()](https://github.com/tarantool/tarantool/blob/284797792771830215d87683e2796db830d47b07/src/box/lua/load_cfg.lua#L10-L22) function and the [synchronized](https://github.com/tarantool/synchronized/) module).

What should work, but I didn't think about it at writting:

* Hot code reload.
</details>

**Comment by Totktonada** [comment]
Related: https://github.com/tarantool/tarantool/discussions/6800.
